core/shared/src/main/scala-2.11-2.12/ChunkLike.scala

```diff
@@ -0,0 +1,187 @@
+ */
+private[zio] trait ChunkLike[+A] extends IndexedSeq[A] with IndexedSeqLike[A, Chunk[A]] { self: Chunk[A] =>
+
+  /**
+   * Returns a filtered, mapped subset of the elements of this chunk.
+   */
+  override final def collect[B, That](pf: PartialFunction[A, B])(implicit bf: CanBuildFrom[Chunk[A], B, That]): That =
+    bf match {
+      case _: ChunkCanBuildFrom[_] => collectChunk(pf)
+      case _                       => super.collect(pf)
+    }
+
+  /**
+   * Returns the concatenation of mapping every element into a new chunk using
+   * the specified function.
+   */
+  override final def flatMap[B, That](
+    f: A => GenTraversableOnce[B]
+  )(implicit bf: CanBuildFrom[Chunk[A], B, That]): That =
+    bf match {
+      case _: ChunkCanBuildFrom[_] => flatMapChunk(f)
+      case _                       => super.flatMap(f)
+    }
+
+  /**
+   * Returns the first index for which the given predicate is satisfied.
+   */
+  override final def indexWhere(f: A => Boolean): Int =
+    indexWhere(f, 0)
+
+  /**
+   * Returns a chunk with the elements mapped by the specified function.
+   */
+  override final def map[B, That](f: A => B)(implicit bf: CanBuildFrom[Chunk[A], B, That]): That =
+    bf match {
+      case _: ChunkCanBuildFrom[_] => mapChunk(f)
+      case _                       => super.map(f)
+    }
+
+  /**
+   * Generates a readable string representation of this chunk using the
+   * specified start, separator, and end strings.
+   */
+  override final def mkString(start: String, sep: String, end: String): String = {
+    val builder = new scala.collection.mutable.StringBuilder()
+
+    builder.append(start)
+
+    var i   = 0
+    val len = self.length
+
+    while (i < len) {
+      if (i != 0) builder.append(sep)
+      builder.append(self(i).toString)
+      i += 1
+    }
+
+    builder.append(end)
+
+    builder.toString
+  }
+
+  /**
+   * Generates a readable string representation of this chunk using the
+   * specified separator string.
+   */
+  override final def mkString(sep: String): String =
+    mkString("", sep, "")
+
+  /**
+   * Generates a readable string representation of this chunk.
+   */
+  override final def mkString: String =
+    mkString("")
+
+  /**
+   * Determines if the chunk is not empty.
+   */
+  override final def nonEmpty: Boolean =
+    length > 0
+
+  /**
+   * The number of elements in the chunk.
+   */
+  override final def size: Int =
+    length
+
+  /**
+   * The implementation of `flatMap` for `Chunk`.
+   */
+  protected final def flatMapChunk[B, That](f: A => GenTraversableOnce[B]): Chunk[B] = {
+    val len                    = self.length
+    var chunks: List[Chunk[B]] = Nil
+
+    var i               = 0
+    var total           = 0
+    var B0: ClassTag[B] = null.asInstanceOf[ClassTag[B]]
+    while (i < len) {
+      val chunk = f(self(i)) match {
+        case chunk: Chunk[B] => chunk
+        case other           => Chunk.fromIterable(other.toList)
```

Comment:
This `toList` will be detrimental to performance. Probably best to do `Chunk.fromArray(other.toArray)` if we need to do that.

Reply:
Will change.

Reply:
We can't call `toArray` here because we don't have a `ClassTag` at this point. May make sense to refactor this so we just iterate over the original collection and either get the class tag from the `Chunk` if it is a chunk or otherwise from the first value of the collection using `Tag.fromValue`.

Reply:
Right. I think if we can only iterate over it once, then the best we can do is use the chunk builder to turn the "iterable once" into a chunk.

OTOH if we match against a few other cases, e.g. Vector, List, etc., we know what collection type we are dealing with and can handle it more specially (e.g. Vector we can preallocate, List cannot be pre-allocated since it's O(n) to determine the size, etc.).

But open question how much of that performance work to do here. I'm fine merging this in now, we can create tickets for performance work.

Reply:
I've got an idea about how we can do this efficiently. Will do a turn on it now.